### PR TITLE
Adds hardcoded offsets for concept demonstration

### PR DESCRIPTION
**Pull request by l4cr0ss**
@jamesleesaunders thank you for the receptive response to my issue. Please review this initial change (it's very initial! lots of console debug statements and hardcoded values!) to get an idea of what I'm trying to do. I will try to figure out how to get a gist that shows up on bl.ocks.org too, but I've never used that before so no promises. Also, I just use VIM so at the end of this I may need your help understanding the coding conventions you'd like me to use in order to make everything merge ready.

I'll leave in the contents of my commit message below. As I state there this an initial commit with some hardcoded values to hopefully explain what I'm trying to do. 

First set of changes:
```javascript
-      return axisDirectionVector.map(function (a) {
-        return scale(t) * a;
+      return axisDirectionVector.map(function (a, i) {
+        if (axis === "xzAxis" && i === 0) {
+          return scale(t) * a + 4;
+        } else if (axis === "zxAxis" && i === 2) {
+          return scale(t) * a + 6;
+        } else {
+          return scale(t) * a;
+        }
```

The first set of changes you encounter were actually added second in order to equalize the padding between the first and last axes in the xzAxis direction. I wasn't able to make this work using d3 methods like `roundRange()` and `paddingOuter()` alone. It may be that I'm doing something wrong, or it may be that the code currently at `src/component/axis.js:98` is missing something.

Second set of changes:
```javascript
-      newText.attr("translation", tickDirectionVector.map(function (d) {
-        return -d * tickPadding;
-      })).append("billboard").attr("axisofrotation", "0 0 0").append("shape").call(makeSolid, "black").append("text").attr("string", tickFormat).append("fontstyle").attr("size", 1.3).attr("family", "SANS").attr("style", "BOLD").attr("justify", "MIDDLE");
+      newText.attr("translation", tickDirectionVector.map(function (d, i, arr) {
+        var r = -d * tickPadding * -42.5;
+        console.log("arr: " + arr + ", i: " + i + ", d: " + d + ", p: " + tickPadding + ", r: " + r);
+        return r;
```
The second set of changes are inside the `newText` translation that you mentioned to me in the issue I opened. There I've hardcoded an offset of `-42.5` which is multiplied by the `tickPadding` in the direction of the tick vector. That's really the core of what I'm interested in. Being able to replace that number with a derived value that will put the label at the opposite end of the axis it marks.

```
This commit also introduces some logging that will be pulled out later.
The idea is that given a series of labels along an axis, there ought to
be a way to control which end of the axis the label is rendered at. This
way the labels can be seen in the default rendering.

This commit also adds offsets to correct for some challenges encountered
in getting the xzAxis to pad ticks with an equal(ish) amount of spacing.
Trying to do this using the d3 roundRange() and paddingOuter() features
didn't behave as expected, although that may be a misunderstanding on
my part.
```

### CODE REVIEW

**Comment by jamesleesaunders** [comment]
Hi @l4cr0ss I love where you are going with this. Thanks for the picture - I now understand and agree it does appear easier to read in certain chart types. 

GitHub is a little confusing in the way Issues and Pull Requests are separate. What I will do is close Issue #59 and we can continue comms here on this PR #60.

I made a few quick tests and noticed a couple of small glitches:
Screenshot 2019-03-09 at 21 02 28
namely the odd mis placed lines on the X and Z axis.  Check the Example files in a browser to see what I mean.

What I would probably like to see is this being an option which is passed in to the axis component so the developer can then choose the labels to be at either end of the axis.

So rather than the `const axis = selection._groups[0][0]["classList"][0];` bit I would expect this to be an option. Perhaps add a getter/setter option called tickPosition()  with options 'start' and  'end' (or perhaps 'front' / 'rear' )?

This tickPosition() option could also be added to axisThreePlane.js (which it would then pass though to axis.js).

Thanks again for your interest, I am so pleased to be working with you on this.

**Comment by l4cr0ss** [comment]
Great! I am likewise happy that there is interest in collaboration. I have been wrangling with a SQL query for the last couple of hours but when I get it sorted I'll start working on an implementation that works as you describe and accepts options as input. 

Would you be able to push those tests you mentioned to the repo for me? Thanks again for the feedback. I'll almost certainly have a few more questions for you as I go through putting this together.

**Comment by jamesleesaunders** [comment]
Cool. Sounds like a tricky SQL query!

The ‘tests’ I describe are already in the repo in the examples folder.
As well as running the standard unit tests I also tend to check each of the examples to make sure they all look correct.

**Comment by jamesleesaunders** [comment]
I’n sorry I have just merged an update to axis.js which now conflicts with your PR, you may need to resolve the conflicts.

**Comment by jamesleesaunders** [comment]
@l4cr0ss To save you the hassle I have resolved the conflicts and rebased this PR so you can continue if you are still up for it. Do let me know if you have any questions. 
All the best,
Jim

**Comment by l4cr0ss** [comment]
Thank you for the rebase. I was watching your progress on the volume slice stuff and was going to ask. I've not stopped thinking about the project, just had a number of other things going on in the day-to-day keeping me busy.

Also, as I've worked with the library to get my bar chart built I've run into some issues, and I'm having trouble figuring out if it's just "user error" in the sense of me misunderstanding some underlying principles of a framework (d3 and/or x3dom) or if it's me trying to do something with the d3-x3dom library that is outside of the intended scope.

**An example** 

As I believe I've mentioned already I'm working with a multi-series bar chart. One of my requirements is that a series of data be able to be filtered dynamically. I've implemented this as a series of checkboxes. If the checkbox is checked, then the data point should be present in the series. 

The trouble then arises as the user interacts with the graph by checking and unchecking boxes to filter and unfilter categories of data. The mechanism by which the tick marks are rendered and transitioned into and out of the graph seems to differ from the mechanism by which the bars do the same, resulting in a bad state where the tick labels and the ticks aren't lined up with the visualization of the data by the bars. 

Now, I've found that I can work around this by creating an entirely new axisThreePlane component composed of ticks with labels describing only those data points in the series that have not been filtered out, and then transitioning this new component into the existing scene along with a `call(data)` and `call(chart)`. However, I'm still left with doubts about whether or not this is the right way to accomplish the task, and as I started out saying I've not had the time to do the digging to understand if it's really just me misunderstanding something in the design. It seems like there is the possibility here of making the tick labels into their own component that can then, in some sense of the word, "decorate" other components.

**tldr;**
All of that to say I haven't stopped thinking about the problem and still have every intention of contributing to the project. I maintain the thought that this particular change will be small and targeted and nothing crazy - really, it should be as you described as straightforward as creating a parameter or set of parameters for toggling the placement of the tick labels - it will just be a little longer before I can turn my attention to getting it done as I need to finish up "getting the thing working" for a client.

**Comment by jamesleesaunders** [comment]
d3-x3dom *should* cope with dynamically addming / removing data (this after all is one of d3’s greatest strengths. I’d be interested to check this though, if you have not got an example already I may put together a CodePen example to try. 
If the axis does not re-scale on data changes this should hopefully be an easy fix (it will just be to do with the D3 enter/update pattern part of the code and maybe a missed call to merge).

I’ll have a play.

In the mean time if you want to continue with the labels at the other end that would be cool.

**Comment by l4cr0ss** [comment]
I will try to whip up an example as well, but imagine you want to hide the label without removing the grid line. That's what I'm having trouble doing, because the labels are glued to the ticks.

**Comment by jamesleesaunders** [comment]
Hi @l4cr0ss I have not forgotten about you. Your comment about being able to add/remove data to refresh a chart has sent me on a mission of fun and games! I thought it would be very easy to make d3-x3dom charts and components reactive to be able to add or remove data series on the fly - this, it  turns out, was not as straight forwards as I thought.

I am still trying to iron out some bugs but these examples are where I am:

Bar Charts (working)
* https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterChartMultiSeriesBar.html
* https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterChartSingleSeriesBar.html
* https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterComponentMultiSeriesBar.html

Bubble Chart (partially working but unable to add new series)
* https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterComponentBubbles.html

Ribbon Chart (partially working but ribbon widths don't always change)
* https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterChartRibbon.html
* https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterComponentRibbon.html

I know this is not exactly what you were doing but thought you may be interested.

I am still totally open to you working on making the position of the Axis labels customisable.

**Comment by l4cr0ss** [comment]
Thanks for the update and for sharing. I have not forgotten either! Just been busy. 

What you've done in the Bar Charts example #1 is actually exactly what I have been working on. I was able to make it work as you have in https://raw.githack.com/jamesleesaunders/d3-x3dom/Transitions/examples/tests/FilterChartMultiSeriesBar.html but it was very hacky. 

I will try to carve out some time tomorrow evening to revisit making the Axis labels customisable because I think this would be a very nice feature.

**Comment by jamesleesaunders** [comment]
Cool! Well that is good to hear my head scratching over the past couple of weeks has not been in vein.
At least the Bar Chart transition is working. I still have some problems getting the Bubble and Ribbon charts working.

I have now merged into master all the transition changes I have made so all changes are in and ready for you to play with should you wish.

All the best,
Jim

**Comment by l4cr0ss** [comment]
@jamesleesaunders Please take a look at the commits I've just pushed when you have the time. Note the bug that I mention in the final commit where I add the example. As written I do not see an easy way to update the position of existing, rendered labels without redrawing everything.

**Comment by jamesleesaunders** [comment]
Cool! I love your idea of 'proximal' and 'distal' positions.
As I have also been sniffing axis.js to get transitions working properly I have merged in your changes and merged with mine (congrats your first GitHub Pull Request has been accepted ;-) ). 

I am loving the new distal label positions! (after seeing them on the bar chart they look so much better). Thanks! 

I have a couple more tweaks to make to axis for transitions so I will try and see if I can get them to change positions with your LabelFilterChartMultiSeriesBar.html example tonight. I'll let you know how I get on.

In other news... I have also got the ribbon and bubble chars transitions working properly! 

Thanks again for your contributions - you would not believe how pleased I am to have some help. Please do feel free to continue making suggestions and/or PRs.

**Comment by jamesleesaunders** [comment]
> and see if I can get them to change positions with your LabelFilterChartMultiSeriesBar.html example tonight.

In the end I didn't get a chance to look at making the labels proximity transitions, and I don't think Ill get any time now as have a busy couple of weeks coming up over, so if you want to take a look please feel free. I am sure it is only a case of tweaking around with a little code I added recently here:

https://github.com/jamesleesaunders/d3-x3dom/blob/15e3a715000341c778b10027e5f8cfbad1640fe8/src/component/axis.js#L156-L164

If you do figure it out please feel free to pop it in a new PR and Ill merge it in :-)

Happy Easter Holidays!

**Comment by l4cr0ss** [comment]
Happy to take a look as soon as I can. By the way I think I messed up the whitespace in that PR that was merged. I was using spaces where the rest of the code appears to be using tabs. Sorry!

**Comment by jamesleesaunders** [comment]
Huzzah! I figured it out! I had a mini brain wave and fixed the distal/proximal label transitions.

See:
https://raw.githack.com/jamesleesaunders/d3-x3dom/master/examples/tests/LabelFilterChartMultiSeriesBar.html

The labels move from proximal to distal nicely now.

No worries about white space, I just ran the code though my IDE auto-tidy to sort it out, so no problem at all.

**Comment by l4cr0ss** [comment]
> Huzzah! I figured it out! I had a mini brain wave and fixed the distal/proximal label transitions.
 They work great! Nicely done 👍  

Also a very nice refactor. I particularly like the pluralization made on lines 120, 123 (of axis.js), etc. My first couple runs through the codebase I got hung up on the assumption there was just a single one of those things.
